File: PL/3/JavascriptOO/OOProject.js
'use strict';
let fs = require("fs");

function DocumentProcessor(builder) {
    this.lFilterer = builder.lFilterer;
    this.cConverter = builder.cConverter;
    this.wFinder = builder.wFinder;
    this.nonABCFilterer = builder.nonABCFilterer;
    this.wCounter = builder.wCounter;

    this.process = function(text) {
        if (this.nonABCFilterer == null) {
            return this.lFilterer.process(text);
        }  else if (this.lFilterer == null) {
            return this.wCounter.process(this.nonABCFilterer
                .process(this.wFinder
                    .process(this.cConverter
                        .process(text))));
        } else {
            return this.wCounter.process(this.nonABCFilterer
                .process(this.wFinder
                    .process(this.cConverter
                        .process(this.lFilterer
                            .process(text)))));
        }
    }
}

function DocumentProcessorBuilder() {

    this.lFilterer = null;
    this.cConverter = null;
    this.wFinder = null;
    this.nonABCFilterer = null;
    this.wCounter = null;

    this.setLineFilterer = function(lf) {
        this.lFilterer = lf;
        return this;
    };

    this.setCConverter = function(cc) {
        this.cConverter = cc;
        return this;
    };

    this.setWFinder = function(wf) {
        this.wFinder = wf;
        return this;
    };

    this.setNonABCFilterer = function(abcf) {
        this.nonABCFilterer = abcf;
        return this;
    };

    this.setWCounter = function(wc) {
        this.wCounter = wc;
        return this;
    };

    this.build = function() {
        return new DocumentProcessor(this);
    };
}

//command objects:

function LineFilterer(toFind) {
    this.toFind = toFind;
    this.process = function(text) {
        let lines = text.split(/\r?\n/);
        let i = lines.length;
        while (i--) {
            if (!lines[i].includes(this.toFind)) {
                lines.splice(i, 1);
            }
        }
        return lines.join('\n');
    };
}

function CaseConverter() {
    this.process = function(text) {
        return text.toLowerCase();
    };
}

function WordFinder() {
    this.process = function(text) {
        return text.split(/(\s+)/).filter( word => word.trim().length > 0 );
    };
}

function NonABCFilterer() {
    this.process = function(words) {
        let result = [];
        for(let i = 0; i < words.length; i++) {
            result[i] = words[i].replace(/[^0-9a-z]/gi, '');
        }
        return result.filter(Boolean);
    };
}

function WordCounter() {
    this.process = function(words) {
        let result = {};
        words.forEach(function (key) {
            if (result.hasOwnProperty(key)) {
                result[key]++;
            } else {
                result[key] = 1;
            }
        });
        return result;
    };
}

//process controller:

function ProcessController() {
    this.parse = function(command) {
        if (command[2] != null) {
            if (command[2] === "wc") {
                let result = this.wc(fs.readFileSync(command[3]).toString('utf-8'));
                for (let key in result) {
                    if (result.hasOwnProperty(key)) {
                        console.log(key + "\t" + result[key]);
                    }
                }
            } else if (command[2] === "grep") {
                if (command.length === 5) {
                    console.log(this.grep(command[3],
                        fs.readFileSync(command[4]).toString('utf-8')));
                } else {
                    let result = this.pipe(command[3],
                        fs.readFileSync(command[4]).toString('utf-8'));
                    for (let key in result) {
                        if (result.hasOwnProperty(key)) {
                            console.log(key + "\t" + result[key]);
                        }
                    }
                }
            } else {
                console.log("Invalid Command.");
            }
        } else {
            console.log("Requires command line args.");
        }
    };

    //command functions
    this.grep = function(searchString, text) {
        return new DocumentProcessorBuilder()
            .setLineFilterer(new LineFilterer(searchString))
            .build().process(text);
    };
    this.wc = function(text) {
        return new DocumentProcessorBuilder()
            .setCConverter(new CaseConverter())
            .setWFinder(new WordFinder())
            .setNonABCFilterer(new NonABCFilterer())
            .setWCounter(new WordCounter())
            .build().process(text);
    };
    this.pipe = function (searchString, text) {
        return new DocumentProcessorBuilder()
            .setLineFilterer(new LineFilterer(searchString))
            .setCConverter(new CaseConverter())
            .setWFinder(new WordFinder())
            .setNonABCFilterer(new NonABCFilterer())
            .setWCounter(new WordCounter())
            .build().process(text);
    };

}

//parse command line args

new ProcessController().parse(process.argv);

File: PL/3/PythonOO/OOProject.py
import sys


class DocumentProcessor(object):
    def __init__(self, builder):
        self.lineFilterer = builder.lineFilterer
        self.caseConverter = builder.caseConverter
        self.nonABCFilterer = builder.nonABCFilterer
        self.wordCounter = builder.wordCounter
        self.wordFinder = builder.wordFinder

    # processes text based on which objects exist:
    def process(self, text):
        if self.nonABCFilterer is None:
            return self.lineFilterer\
                .process(text)
        elif self.lineFilterer is None:
            return self.wordCounter\
                .process(self.nonABCFilterer
                        .process(self.wordFinder
                                 .process(self.caseConverter
                                          .process(text))))
        else:
            return self.wordCounter\
                .process(self.nonABCFilterer
                         .process(self.wordFinder
                                  .process(self.caseConverter
                                           .process(self.lineFilterer
                                                    .process(text)))))


class DocumentProcessorBuilder:

    def build(self):
        return DocumentProcessor(self)

    def __init__(self):
        self.lineFilterer = None
        self.caseConverter = None
        self.nonABCFilterer = None
        self.wordCounter = None
        self.wordFinder = None

    def set_line_filterer(self, linefilterer):
        self.lineFilterer = linefilterer
        return self

    def set_case_converter(self, caseconverter):
        self.caseConverter = caseconverter
        return self

    def set_non_abc_filterer(self, nonabcfilterer):
        self.nonABCFilterer = nonabcfilterer
        return self

    def set_word_counter(self, wordcounter):
        self.wordCounter = wordcounter
        return self

    def set_word_finder(self, wordfinder):
        self.wordFinder = wordfinder
        return self


class LineFilterer:
    # the text to find. will be initialized with class initialization
    toFind = None

    def __init__(self, tofind):
        LineFilterer.toFind = tofind

    @staticmethod
    def process(text):
        lines = [line for line in text.splitlines() if line.find(LineFilterer.toFind) != -1]
        return '\n'.join(lines)


class CaseConverter:
    @staticmethod
    def process(text):
        return text.lower()


class WordFinder:
    @staticmethod
    def process(text):
        return text.split()


class NonABCFilterer:
    @staticmethod
    def process(words):
        result = []
        for word in words:
            result.append("".join([c for c in word if c.isalnum()]))
        return list(filter(None, result))


class WordCounter:
    @staticmethod
    def process(words):
        uniquewords = set(words)
        result = {}
        for word in uniquewords:
            result[word] = words.count(word)
        return result


class ProcessController(object):

        @staticmethod
        def grep(searchstring, text):
            return DocumentProcessorBuilder().set_line_filterer(LineFilterer(searchstring)) \
                .build().process(text)

        @staticmethod
        def pipe(searchstring, text):
            return DocumentProcessorBuilder()\
                .set_line_filterer(LineFilterer(searchstring))\
                .set_case_converter(CaseConverter())\
                .set_word_finder(WordFinder)\
                .set_non_abc_filterer(NonABCFilterer())\
                .set_word_counter(WordCounter)\
                .build().process(text)

        @staticmethod
        def wc(text):
            return DocumentProcessorBuilder()\
                .set_case_converter(CaseConverter())\
                .set_word_finder(WordFinder)\
                .set_non_abc_filterer(NonABCFilterer())\
                .set_word_counter(WordCounter)\
                .build().process(text)

        @staticmethod
        def parse(command):
            if len(command) > 1:
                if command[1] == 'wc':
                    file = command[2]
                    with open(file, 'r') as myfile:
                        text = myfile.read()
                        count = ProcessController.wc(text)
                        for word, freq in count.items():
                            print(word + '\t' + str(freq))
                elif command[1] == 'grep':
                    file = command[3]
                    with open(file, 'r') as myfile:
                        text = myfile.read()
                        if len(command) == 4:
                            print(ProcessController.grep(str(command[2]), text))
                        else:
                            count = ProcessController.pipe(str(command[2]), text)
                            for word, freq in count.items():
                                print(word + '\t' + str(freq))
                else:
                    print('Invalid Command.')
            else:
                print('Requires command line args.')


ProcessController().parse(sys.argv)
